spec/models/partnership_csv_upload_spec.rb

```diff
@@ -33,4 +38,106 @@
+    it "finds invalid URNs" do
+      given_the_csv_contains_urns([valid_school.urn, 1])
+
+      expect(@subject.invalid_schools.length).to eql 1
+      expect(@subject.invalid_schools).to contain_exactly({ urn: "1", row_number: 2, school_name: "", message: "URN is not valid" })
+    end
+
+    it "finds CIP only schools" do
+      given_the_csv_contains_urns([welsh_school.urn])
+
+      expect(@subject.invalid_schools.length).to eql 1
+      expect(@subject.invalid_schools).to contain_exactly(
+        {
+          urn: welsh_school.urn,
+          row_number: 1,
+          school_name: welsh_school.name,
+          message: "School not eligible for funding",
+        },
+      )
+    end
+
+    it "finds ineligible schools" do
+      given_the_csv_contains_urns([closed_school.urn])
+
+      expect(@subject.invalid_schools.length).to eql 1
+      expect(@subject.invalid_schools).to contain_exactly(
+        {
+          urn: closed_school.urn,
+          row_number: 1,
+          school_name: closed_school.name,
+          message: "School not eligible for inductions",
+        },
+      )
+    end
+
+    it "finds schools already in a partnership with the lead provider" do
+      partnered_school = create(:school)
+      given_the_csv_contains_urns([partnered_school.urn])
+      Partnership.create!(
+        school: partnered_school,
+        lead_provider: @subject.lead_provider,
+        delivery_partner: @subject.delivery_partner,
+        cohort: current_cohort,
+      )
+
+      expect(@subject.invalid_schools.length).to eql 1
+      expect(@subject.invalid_schools).to contain_exactly(
+        {
+          urn: partnered_school.urn,
+          row_number: 1,
+          school_name: partnered_school.name,
+          message: "Your school - already confirmed",
+        },
+      )
+    end
+
+    it "finds schools already in a partnership with a different lead provider" do
+      partnership = create(:partnership, cohort: current_cohort)
+      partnered_school = partnership.school
+      given_the_csv_contains_urns([partnered_school.urn])
+
+      expect(@subject.invalid_schools.length).to eql 1
+      expect(@subject.invalid_schools).to contain_exactly(
+        {
+          urn: partnered_school.urn,
+          row_number: 1,
+          school_name: partnered_school.name,
+          message: "Recruited by other provider",
+        },
+      )
+    end
+  end
+
+  describe "#urns" do
+    it "returns the correct urns" do
+      urns = 5.times.map { Faker::Number.unique.decimal_part(digits: 7).to_s }
+      given_the_csv_contains_urns(urns)
+
+      expect(@subject.urns).to eql urns
+    end
+  end
+
+  describe "#valid_schools" do
+    it "returns only valid schools" do
+      given_the_csv_contains_urns([valid_school.urn, closed_school.urn])
+
+      expect(@subject.valid_schools).to contain_exactly(valid_school)
+    end
+  end
+
+private
+
+  def given_the_csv_contains_urns(urns)
+    file = Tempfile.new
```

Comment:
Any reason to not use a fixture rather than a Tempfile?

Reply:
Because it's easier to use generated schools than to hardcode it all